pkg/providers/ibmcloud.go

```diff
@@ -0,0 +1,331 @@
+package providers
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/softlayer/softlayer-go/datatypes"
+	"github.com/softlayer/softlayer-go/services"
+	"github.com/softlayer/softlayer-go/session"
+	"github.com/softlayer/softlayer-go/sl"
+
+	"github.com/openshift/ofcir/pkg/utils"
+)
+
+type ibmcloudProviderConfig struct {
+	APIKey string `json:"apikey"`
+	Sshkey string `json:"sshkey"`
+	Preset string `json:"preset"`
+}
+
+type ibmcloudProvider struct {
+	config ibmcloudProviderConfig
+	client *session.Session
+}
+
+func IbmcloudProviderFactory(providerInfo string, secretData map[string][]byte) (Provider, error) {
+	config := ibmcloudProviderConfig{}
+
+	if configJSON, ok := secretData["config"]; ok {
+		if err := json.Unmarshal(configJSON, &config); err != nil {
+			return nil, fmt.Errorf("error in provider config json: %w", err)
+		}
+	}
+
+	provider := &ibmcloudProvider{
+		config: config,
+	}
+
+	provider.client = session.New("apikey", provider.config.APIKey)
+	return provider, nil
+}
+
+func (p *ibmcloudProvider) getPackageItems(packageId int) (resp []datatypes.Product_Item) {
+	var mask = "id, itemCategory, keyName,prices[id, hourlyRecurringFee, recurringFee, categories]"
+	var service = services.GetProductPackageService(p.client)
+	receipt, err := service.Id(packageId).Mask(mask).GetItems()
+	if err != nil {
+		return
+	}
+
+	return receipt
+}
+
+func (p *ibmcloudProvider) getStandardPrice(item datatypes.Product_Item) (resp datatypes.Product_Item_Price) {
+	for _, itemPrice := range item.Prices {
+		if itemPrice.LocationGroupId == nil {
+			return itemPrice
+		}
+	}
+	return datatypes.Product_Item_Price{}
+
+}
+
+func (p *ibmcloudProvider) getItemPriceList(packageId int, itemKeyNames []string) (resp []datatypes.Product_Item_Price) {
+
+	items := p.getPackageItems(packageId)
+	var prices []datatypes.Product_Item_Price
+
+	for _, itemKeyName := range itemKeyNames {
+		for _, item := range items {
+			if (*item.KeyName) == itemKeyName {
+				itemPrice := p.getStandardPrice(item)
+				prices = append(prices, itemPrice)
+				break
+			}
+		}
+	}
+
+	return prices
+}
+
+func (p *ibmcloudProvider) serverCreate(pool, presetname string) (string, error) {
+
+	// TODO: Would love to find a way to list locations where the package is available but
+	// I can't for the life of me figure it out, need to revisit
+	// I've removed some of the more expensive locations from the list below
+	locations := []string{"DALLAS09", "DALLAS10", "DALLAS12", "DALLAS13", "FRANKFURT", "FRANKFURT04", "FRANKFURT05", "LONDON02", "LONDON04",
```

Comment:
This could be a const either. Do we really need to shuffle amongst all these locations? Maybe we can just keep a bunch of them

Reply:
> This could be a const either.
Will move to CONST,

> Do we really need to shuffle amongst all these locations?
we are only try once per reconcile loop, if it doesn't work a shuffle ensures we don't keep trying the same one, the alternative would be to try multiple calls in the same reconcile loop

> Maybe we can just keep a bunch of them
I've already remove some of them, If we know which we most reliable we could use those but right know I don't know which are best...